api/profile/src/model.rs

```diff
@@ -0,0 +1,125 @@
+use std::collections::HashMap;
+
+use anyhow::{anyhow, Result, Ok};
+use bytes::Bytes;
+use http::HeaderMap;
+use serde::{Deserialize, Serialize};
+
+use spin_sdk::pg::{self as db, Decode, ParameterValue, Row};
+
+use crate::utils::{get_last_param_from_route, get_column_lookup};
+
+fn as_param<'a>(value: &'a Option<String>) -> Option<ParameterValue<'a>> {
+    match value {
+        Some(value) => Some(ParameterValue::Str(value.as_str())),
+        None => None
+    }
+}
+
+fn as_nullable_param<'a>(value: &'a Option<String>) -> ParameterValue<'a> {
+    match as_param(value) {
+        Some(value) => value,
+        None => ParameterValue::DbNull,
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub(crate) struct Profile {
+    pub id: Option<String>,
+    pub handle: String,
+    pub avatar: Option<String>,
+}
+
+impl Profile {
+    pub(crate) fn from_path(headers: &HeaderMap) -> Result<Self> {
+        let header = headers.get("spin-path-info").ok_or(anyhow!("Error: Failed to discover path"))?;
+        let path = header.to_str()?;
+        match get_last_param_from_route(path) {
+            Some(handle) => Ok(Profile {
+                id: None,
+                handle: handle,
+                avatar: None,
+            }),
+            None => Err(anyhow!("Failed to parse handle from path")),
+        }
+    }
+
+    pub(crate) fn from_bytes(b: &Bytes) -> Result<Self> {
+        Ok(serde_json::from_slice(&b)?)
+    }
+
+    fn from_row(row: &Row, columns: &HashMap<&str, usize>) -> Result<Self> {
+        let id = String::decode(&row[columns["id"]]).ok();
+        let handle = String::decode(&row[columns["handle"]])?;
+        let avatar = String::decode(&row[columns["avatar"]]).ok();
+        Ok(Profile {
+            id,
+            handle,
+            avatar,
+        })
+    }
+
+    pub(crate) fn insert(&self, db_url: &str) -> Result<()> {
+        let params = vec![
+            as_param(&self.id).ok_or(anyhow!("The id field is currently required for insert"))?,
+            ParameterValue::Str(&self.handle),
+            match as_param(&self.avatar) {
+                Some(p) => p,
+                None => ParameterValue::DbNull,
+            }
+        ];
+        db::execute(db_url, "INSERT INTO profiles (id, handle, avatar) VALUES ($1, $2, $3)", &params)?;
```

Comment:
Glancing at the `pg` sdk error handling, it seems like we'd want `.context("executing database statement")?` or something more descriptive here.

Reply:
I have seen that used in places but wasn't sure what it actually does. Still figuring out how to use the `Result` and `Option` types properly.